### ISSUE TITLE
Alternatives to define_private_input?

### ISSUE THREAD

**Issue by tshead2** [mortendahl/tf-encrypted#395]
I'm setting-up some experiments that involve evaluating observations using an already-trained model loaded from disk.  I'm curious whether there's any way to run code on the model-provider without using define_private_input()?  Ideally, I'd like the model-provider and input-provider to each be able to run their own private scripts, instead of having to reveal details like filesystem paths to a single centralized script, as is done in the current examples.

Thanks in advance!
Tim

**Comment by tshead2** [comment]
This runs even deeper than I realized at first (I admit that I'm inexperienced with distributed TF).  The callback you pass to define_private_input() is executed on the master, not the (in this case) model-provider, is that correct?  Which means that I must use TF functions to load my data.  I'd suggest rethinking the API, as passing a callback to define_private_input() that is immediately called seems somewhat misleading.

**Comment by justin1121** [comment]
Hello Tim,

Thanks for your interest and your issue!

Something to understand about tensorflow distributed is that there is one script that the master executes. This script builds a tensorflow computation graph which includes delegating certain operations to the correct parties. For example, when running a matrix multiplication, some operations are run on the crypto provider, server 0 and server1.

We currently do not have a way to run separate scripts on different parties. It is quite possible that we could run code on another party without the master knowing about it but we don't really have a design or path forward for this. Suggestions welcome!

As for your concern with the API. You can see from this line https://github.com/mortendahl/tf-encrypted/blob/master/tf_encrypted/protocol/pond.py#L452, tf.device pins the inputter_fn to whatever device is specified in the input to `define_private_input`. Note that if the weights are hardcoded in the inputter function then these weights would be leaked from the master to the weight-provider for example. But, you could load the weights from disk using some tensorflow functions. I think you've realized already that you would be leaking some information about the weights provider. Let us know if you have any suggestions on how this could be improved.

Thanks again,

Justin

**Comment by tshead2** [comment]
Yeah, I'm just unhappy at having to shoehorn my data (a pre-trained copy of VGG-16) into Tensorflow's data API.  Coming from an MPI background, I'm accustomed to having the flexibility to run any code I want on any process in the cluster.  Obviously, this is a limitation of Tensorflow's distributed model, and not your fault :)

**Comment by justin1121** [comment]
Understandable. Even though its not our fault its valuable feedback and something for us to consider going forward!

**Comment by mortendahl** [comment]
Thanks for the feedback @tshead2!

To add a quick note regarding your first question, it is in fact possible to run several independent scripts instead of only a single one, yet it's not something we've focused on so far. TensorFlow calls it [between-graph replication](https://www.tensorflow.org/deploy/distributed) and I've posted a small example here: https://github.com/dropoutlabs/research/tree/master/tf-between-graph.

Supporting between-graph replicated might also allow us to get around the second issue you raise about getting data into the system. While using the data api is the recommended practice from TensorFlow's side, I agree that it would be convenient to execute arbitrary code on other machines via e.g. [`py_func`](https://www.tensorflow.org/api_docs/python/tf/py_func). However, that (currently) isn't supported from TensorFlow's side and not too optimistic that it will be in the near future.
